Sources/Internal/Logger/DataConnectLogger.swift

```diff
@@ -0,0 +1,94 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import FirebaseCore
+import OSLog
+
+let privateLogDisabledArgument = "-FIRPrivateLogDisabled"
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+class DataConnectLogger {
+  static let logger = Logger(
+    subsystem: "com.google.firebase",
+    category: "[FirebaseDataConnect]"
+  )
+
+  static let privateLoggingEnabled: Bool = {
+    let arguments = ProcessInfo.processInfo.arguments
+    if arguments.contains(privateLogDisabledArgument) {
+      DataConnectLogger.debug("DataConnect private logging disabled.")
+      return false
+    } else {
+      DataConnectLogger.debug("DataConnect private logging enabled.")
+      return true
+    }
+  }()
+
+  private static let logPrefix = "\(Version.sdkVersion) - [FirebaseDataConnect]"
+
+  static var logLevel: FirebaseLoggerLevel {
+    return FirebaseConfiguration.shared.loggerLevel()
+  }
+
+  static func error(_ message: String, code: MessageCode = .placeHolder) {
+    if logLevel.rawValue >= FirebaseLoggerLevel.error.rawValue {
+      let messageCode = String(format: "I-FDC%06d", code.rawValue)
+      logger.error("\(logPrefix)[\(messageCode)] \(message)")
+    }
+  }
+
+  static func warning(_ message: String, code: MessageCode = .placeHolder) {
+    if logLevel.rawValue >= FirebaseLoggerLevel.warning.rawValue {
+      let messageCode = String(format: "I-FDC%06d", code.rawValue)
+      logger.warning("\(logPrefix)[\(messageCode)] \(message)")
+    }
+  }
+
+  static func notice(_ message: String, code: MessageCode = .placeHolder) {
+    if logLevel.rawValue >= FirebaseLoggerLevel.notice.rawValue {
+      let messageCode = String(format: "I-FDC%06d", code.rawValue)
+      logger.notice("\(logPrefix)[\(messageCode)] \(message)")
+    }
+  }
+
+  static func info(_ message: String, code: MessageCode = .placeHolder) {
+    if logLevel.rawValue >= FirebaseLoggerLevel.info.rawValue {
+      let messageCode = String(format: "I-FDC%06d", code.rawValue)
+      logger.info("\(logPrefix)[\(messageCode)] \(message)")
+    }
+  }
+
+  static func debug(_ message: String, code: MessageCode = .placeHolder) {
+    if logLevel.rawValue >= FirebaseLoggerLevel.debug.rawValue {
+      let messageCode = String(format: "I-FDC%06d", code.rawValue)
+      logger.debug("\(logPrefix)[\(messageCode)] \(message)")
+    }
+  }
+}
+
+enum LogPrivacy {
+  case `public`
+  case `private`
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+extension DefaultStringInterpolation {
+  mutating func appendInterpolation(_ value: String, privacy: LogPrivacy = .public) {
+    if privacy == .private, DataConnectLogger.privateLoggingEnabled {
+      appendLiteral(" <private>")
```

Comment:
nit: I think the leading whitespace makes the usage a little harder to understand.

```suggestion
      appendLiteral("<private>")
```

In the below example, my understanding is that `Error: foo when creating GrpcClient <private>"` should be printed. If so, I think it's reasonable to let the caller add the leading space instead of doing it in this method. Example:

```diff
- "Error:\(error) when creating GrpcClient\(self.description, privacy: .private)."` 
+ "Error:\(error) when creating GrpcClient \(self.description, privacy: .private)."` 
```